### PR TITLE
Add auto-detection feature to `esp-println`

### LINKED ISSUE #1576: esp-println: Add `auto-detect` feature
In the flasher stub we have some code to detect (from ROM code) which transport method is connected: https://github.com/esp-rs/esp-flasher-stub/blob/e9b61293aa181a6ef8ffb326ce4360cf2f1162c3/src/lib.rs#L43-L77.

Maybe we can adapt this to use in esp-println, so that no output method feature is required. It might help in cases such as https://github.com/esp-rs/esp-wifi/pull/452.

One thing I'm not so sure about is when this detection should occur, I think my initial instinct would be to make it lazy and check on first access and then keep using that method until reset.

### PR DESCRIPTION
**Pull request by bjoernQ**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
Fixes #1576 

Using the ROM function didn't work for me - `buf_uart_no` was always 0 for me but in general I think using SOF-interrupt-flag is a better way to do what we need here.

I also added a CHANGELOG.md for `esp-println` (but no CI checks for that in this PR)

#### Testing
Running the examples which use `esp-println` should print to UART or Serial-JTAG without reconfiguration (or re-compilation).

### CODE REVIEW

File: esp-println/README.md [hunk 1]
```diff
@@ -39,8 +39,9 @@ You can now `println!("Hello world")` as usual.
   `esp32c3`, `esp32c6`, `esp32h2`, `esp32s2`, and `esp32s3`.
   - One of these features must be enabled.
   - Only one of these features can be enabled at a time.
-- There is one feature for each supported communication method: `uart`, `jtag-serial` and `no-op`.
+- There is one feature for each supported communication method: `uart`, `jtag-serial` and `auto`.
   - Only one of these features can be enabled at a time.
+- `no-op`: don't print anything
```
Comment (JurajSadel) [inline esp-println/README.md:44]: Sorry for nitpicking, we are not consistent here with upper-case/lower-case and colons.

**Review by JurajSadel — APPROVED** [review]
LGTM, thanks!
I left a very nitpick comment (sorry) I noticed while reading the docs. Feel free to ignore it.

**Review by MabezDev — CHANGES_REQUESTED** [review]
Thank you! I think checking the SoF interrupt is much nicer than the ROM function (I think the main reason I went with the ROM function in the stub is that eventually we will also have to detect USB OTG too).

Thanks for adding the changelog too. Would you mind adding esp-println to CI too? After that I think this is ready to go :).

**Review by MabezDev — APPROVED** [review]
Thank you!

**Comment by snoweuph** [comment]
Maybe a stupid necroposting question, but the changelog pointed to this pull:

What does auto detection have to do with removing support for 8266 ?

**Comment by bugadani** [comment]
Nothing at all. You can see that 8266 removal is placed under a version that has already been released before this PR.

**Comment by snoweuph** [comment]
ahh okay, thanks.

**Pull request merged**